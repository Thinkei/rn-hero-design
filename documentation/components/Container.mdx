---
name: Container
menu: Components
---

import { Props } from 'gatsby-theme-docz/src/components/Props';

# Container

## Properties

<Props props_={{
  children: { type: 'Element', required: true },
  fluid: { type: 'Boolean', required: false, defaultValue: { value: 'false' } },
  direction: { type: 'String', required: false },
  style: { type: 'Object', required: false },
  theme: { type: 'Object', required: false, description: 'The theme object contains style of all components. Can be injected by `injectTheme` HOC' },
}} />

