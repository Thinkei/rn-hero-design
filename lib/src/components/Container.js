import React from 'react';
import t from 'prop-types';
import { View, StyleSheet } from 'react-native';
import injectTheme from '../helpers/injectTheme';

const Container = ({
  children,
  fluid = false,
  direction = 'column',
  style,
  theme,
}) => (
  <View
    style={StyleSheet.flatten([
      theme.container.container,
      fluid ? theme.container.fluid : null,
      { flexDirection: direction },
      style,
    ])}
  >
    {children}
  </View>
);

Container.propTypes = {
  children: t.oneOfType([t.element, t.arrayOf(t.element)]),
  fluid: t.bool,
  direction: t.oneOf(['column', 'row']),
  style: t.object,
  /**
   * The theme object contains style of all components. Can be injected by `injectTheme` HOC
   */
  theme: t.object,
};

export default injectTheme(Container);
