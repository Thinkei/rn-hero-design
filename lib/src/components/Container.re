open ReactNative;

[@genType]
[@bs.deriving jsConverter]
type containerDirection = [ | `column | `row | `columnReverse | `rowReverse];

let noStyle = Style.style();

let (|?) = (x, y) =>
  switch (x) {
  | Some(x) => x
  | None => y
  };

[@genType]
[@react.component]
let make =
    (
      ~children,
      ~fluid=false,
      ~direction: option(containerDirection)=?,
      ~style=?,
      ~theme=Hero_Theme.default,
    ) => {
  let direction = direction->Belt.Option.getWithDefault(`column);
  <View
    style={StyleSheet.flatten([|
      theme##container##container,
      fluid ? theme##container##fluid : noStyle,
      Style.style(~flexDirection=direction, ()),
      style |? noStyle,
    |])}>
    children
  </View>;
};

[@genType]
let default = Helpers.injectTheme(make);
