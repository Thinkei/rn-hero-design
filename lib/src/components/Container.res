open ReactNative

@deriving(jsConverter)
type containerDirection = [#column | #row | #columnReverse | #rowReverse]

let noStyle = Style.style()

let \"|?" = (x, y) =>
  switch x {
  | Some(x) => x
  | None => y
  }

@react.component
let make = (~children, ~fluid=false, ~direction, ~style=?, ~theme=Hero_Theme.default) => {
  let direction = containerDirectionFromJs(direction)->Belt.Option.getWithDefault(#column)
  <View
    style={StyleSheet.flatten([
      theme["container"]["container"],
      fluid ? theme["container"]["fluid"] : noStyle,
      Style.style(~flexDirection=direction, ()),
      \"|?"(style, noStyle),
    ])}>
    children
  </View>
}

let default = Helpers.injectTheme(make)
