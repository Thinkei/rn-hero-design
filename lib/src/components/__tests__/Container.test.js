import React from 'react';
import Container from '../Container.gen';
import renderer from 'react-test-renderer';

describe('Container', () => {
  it.each`
    element
    ${<Container direction="column" fluid="true" style={{ backgroundColor: 'yellow' }} />}
    ${<Container direction="row" fluid="false" style={{ backgroundColor: 'yellow' }} />}
    ${<Container direction="rowReverse" fluid="true" style={{ backgroundColor: 'yellow' }} />}
    ${<Container direction="columnReverse" fluid="true" style={{ backgroundColor: 'yellow' }} />}
    ${<Container />}
  `('renders correctly', ({ element }) => {
    const tree = renderer.create(element).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
