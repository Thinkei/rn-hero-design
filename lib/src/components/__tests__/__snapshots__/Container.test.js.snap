// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "yellow",
      "flex": 1,
      "flexDirection": "column",
      "padding": 16.25,
    }
  }
/>
`;

exports[`Container renders correctly 2`] = `
<View
  style={
    Object {
      "backgroundColor": "yellow",
      "flex": 1,
      "flexDirection": "row",
      "padding": 16.25,
    }
  }
/>
`;

exports[`Container renders correctly 3`] = `
<View
  style={
    Object {
      "backgroundColor": "yellow",
      "flex": 1,
      "flexDirection": "row-reverse",
      "padding": 16.25,
    }
  }
/>
`;

exports[`Container renders correctly 4`] = `
<View
  style={
    Object {
      "backgroundColor": "yellow",
      "flex": 1,
      "flexDirection": "column-reverse",
      "padding": 16.25,
    }
  }
/>
`;

exports[`Container renders correctly 5`] = `
<View
  style={
    Object {
      "flexDirection": "column",
      "padding": 16.25,
    }
  }
/>
`;

exports[`Container renders correctly expect flexDirection is column 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "yellow",
        "flex": 1,
        "flexDirection": "column",
        "padding": 16.25,
      }
    }
  />
</View>
`;

exports[`Container renders correctly expect flexDirection is column 2`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "flexDirection": "column",
        "padding": 16.25,
      }
    }
  />
</View>
`;

exports[`Container renders correctly expect flexDirection is column-reverse 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "yellow",
        "flex": 1,
        "flexDirection": "column-reverse",
        "padding": 16.25,
      }
    }
  />
</View>
`;

exports[`Container renders correctly expect flexDirection is row 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "yellow",
        "flex": 1,
        "flexDirection": "row",
        "padding": 16.25,
      }
    }
  />
</View>
`;

exports[`Container renders correctly expect flexDirection is row-reverse 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "yellow",
        "flex": 1,
        "flexDirection": "row-reverse",
        "padding": 16.25,
      }
    }
  />
</View>
`;
